pkg/controllers/data.go

```diff
@@ -0,0 +1,147 @@
+package controllers
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	. "github.com/CESARBR/knot-cloud-storage/pkg/entities"
+	"github.com/CESARBR/knot-cloud-storage/pkg/interactor"
+	"github.com/gorilla/mux"
+)
+
+type DataController struct {
+	DataInteractor interactor.DataInteractor
+}
+
+type DataInterface interface {
+	GetAll()
+	GetByID()
+	Create()
+}
+
+type errorMessage struct {
+	error   bool
+	message string
+}
+
+func NewDataController(dataInteractor interactor.DataInteractor) DataController {
+	return DataController{dataInteractor}
+}
+
+func respondWithError(w http.ResponseWriter, code int, msg string) {
+	respondWithJson(w, code, map[string]string{"message": msg})
+}
+
+func respondWithJson(w http.ResponseWriter, code int, payload interface{}) {
+	response, _ := json.Marshal(payload)
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(code)
+	w.Write(response)
+}
+
+func (d *DataController) GetAll(w http.ResponseWriter, r *http.Request) {
+
+	order, skip, take, startDate, finishDate, errUrl := getUrlQueryParams(r)
+	if errUrl.error != false {
+		respondWithError(w, http.StatusUnprocessableEntity, errUrl.message)
+		return
+	}
+
+	things, err := d.DataInteractor.GetAll(order, skip, take, startDate, finishDate)
+	if err != nil {
+		respondWithError(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	respondWithJson(w, http.StatusOK, things)
+}
+
+func (d *DataController) GetByID(w http.ResponseWriter, r *http.Request) {
+	params := mux.Vars(r)
+	order, skip, take, startDate, finishDate, errUrl := getUrlQueryParams(r)
+	if errUrl.error != false {
+		respondWithError(w, http.StatusUnprocessableEntity, errUrl.message)
+		return
+	}
+	thing, err := d.DataInteractor.GetByID(params["id"], order, skip, take, startDate, finishDate)
+	if err != nil {
+		respondWithError(w, http.StatusBadRequest, "Invalid Thing ID")
+	}
+	respondWithJson(w, http.StatusOK, thing)
+}
+
+func (d *DataController) Save(w http.ResponseWriter, r *http.Request) {
+	var thing Data
+	if err := json.NewDecoder(r.Body).Decode(&thing); err != nil {
+		respondWithError(w, http.StatusBadRequest, "Invalid request payload")
+		return
+	}
+	// thing.ID = bson.NewObjectId()
+	thing.Timestamp = time.Now()
+	if err := d.DataInteractor.Save(thing); err != nil {
+		respondWithError(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	respondWithJson(w, http.StatusCreated, thing)
+	defer r.Body.Close()
+
```

Comment:
Ops, this blank line shouldn't be here :joy:

Reply:
Done!